### PR TITLE
borrowed extractors

### PR DESCRIPTION
**Pull request by aliemjay**
This may be a hacky solution, so I understand if this get rejected for maintainability considerations. However I tried to add comments for the workarounds used here.

### CODE REVIEW

**Comment by aliemjay** [comment]
with "take 2", FromRequest is more semantically correct: from_request returns Self again and Type have a separate type constructor function

**Comment by fakeshadow** [comment]
Thanks for the PR. It looks in good shape already. Let me know when it's ready to be reviewed.

**Comment by fakeshadow** [comment]
LOL. Sorry about that. I was not thinking when deleting the branch

**Comment by aliemjay** [comment]
Ready for review!

Lifetimes may need a more consistent naming but  I'm not sure what to name them.

**Comment by aliemjay** [comment]
take 2 commit is more semantically correct but it add much to the lifetime mess, so I'm not sure if it is worth it.

Also, RefCell needs to  be limited to the payload, otherwise we practically can't have borrowed extractors other than &RefCell<Req>>.

File: http/src/util/service/handler.rs [hunk 1]
```diff
@@ -144,37 +173,45 @@ from_req_impl! { Extract7; A, B, C, D, E, F, G }
 from_req_impl! { Extract8; A, B, C, D, E, F, G, H }
 from_req_impl! { Extract9; A, B, C, D, E, F, G, H, I }
 
-pub trait Handler<Arg>: Clone {
+/// Same as `std::ops::Fn` trait but for async output.
+///
+/// It is necessary in the the HRTB bounds for async fn's with reference paramters because it
+/// allows the output future to be bound to the paramter lifetime.
+///     `F: for<'a> AsyncFn<(&'a u8,) Output=u8>`
+pub trait AsyncFn<Arg> {
```
Comment (aliemjay) [inline http/src/util/service/handler.rs:181]: Separated to be reused elsewhere when needed. I'm thinking of tranform_fn if you like @fakeshadow .

Comment (fakeshadow) [reply http/src/util/service/handler.rs:181]: It's fine. I don't mind move it to service crate or even duplicate it there.

**Comment by fakeshadow** [comment]
Great job. This is working better than I expected.
```rust
use std::{
    cell::{Ref, RefCell},
    convert::Infallible,
    future::{Future},
};

use xitca_http::{
    body::{ResponseBody, RequestBody},
    http::{Request, Response},
    util::service::{get, handler_service, FromRequest, Router},
    HttpServiceBuilder,
};

#[tokio::main]
async fn main() -> std::io::Result<()> {
    xitca_server::Builder::new()
        .bind("test", "localhost:8080", || {
            let router = Router::new()
                .insert("/", get(handler_service(index)));
            HttpServiceBuilder::new(router)
        })?
        .build()
        .await
}

async fn index(
    Extract(path): Extract<Ref<'_, str>>,
) -> Response<ResponseBody> {
    assert_eq!(&*path, "/");
    Response::new("Hello,World!".into())
}

struct Extract<T>(T);

impl<'a> FromRequest<'a, Request<RequestBody>> for Extract<Ref<'a, str>> {
    type Type<'f> = Extract<Ref<'f, str>>;
    type Error = Infallible;
    type Future = impl Future<Output = Result<Self, Self::Error>>;

    fn from_request(req: &'a RefCell<Request<RequestBody>>) -> Self::Future {
        async move {
            let uri = Ref::map(req.borrow(), |req| req.uri().path());
            Ok(Extract(uri))
        }
    }
}
```

**Comment by fakeshadow** [comment]
> take 2 commit is more semantically correct but it add much to the lifetime mess, so I'm not sure if it is worth it.
> 
> Also, RefCell needs to be limited to the payload, otherwise we practically can't have borrowed extractors other than &RefCell<Req>.

I need to think about this. The current code does not limit the input to a sepcific Request type as Req can be anything.
I would if we can use Ref and RefMut to get around some of the limitation.

**Comment by aliemjay** [comment]
> I wonder if we can use Ref and RefMut to get around some of the limitation.

If we don't need mutable access to poll payload, then this is fine, I think.

btw, why the ci is so slow?

File: http/src/util/service/handler.rs (hunk 1 above)
Comment (aliemjay) [reply http/src/util/service/handler.rs:181]: Ok. then I'll try the same thing for transform_fn to remove the cloneability constraint.

**Comment by fakeshadow** [comment]
> > I wonder if we can use Ref and RefMut to get around some of the limitation.
> 
> If we don't need mutable access to poll payload, then this is fine, I think.

My concern is that if there is no internal mutability we lose the ability to mutate the request head completely in extractor and handler. 
There is possible use case that a mutation of the head is wanted and we should also consider the future potential pattern of 
```rust
let req = &mut req;
if let Err(e) = service.call(req).await {
    // the error may mutated the request so we can have a context for handling error.
}
```
> 
> btw, why the ci is so slow?

There are too many features across this repo to keep the dep tree minimal for given specific use case. `cargo-hack` is used to test all the combination of these features to make sure they compile fine without conflict between features. So is for the tests.

**Comment by aliemjay** [comment]
> My concern is that if there is no internal mutability we lose the ability to mutate the request head completely in extractor and handler.

Practically we now can't mutate request in extractor and handler because the RefCell is most likely locked by another extractor. So, I think RefCell is redundant now. 

As for Service<&mut Request>, it cause me headaches just by thinking about it in this context, lol.  Mutable references are hard to reason about here because of the contravariance thing.

**Comment by aliemjay** [comment]
> There are too many features across this repo to keep the dep tree minimal for given specific use case. `cargo-hack` is used to test all the combination of these features to make sure they compile fine without conflict between features. So is for the tests.

despite all of this, the good old `cargo check` for the repo failed because of mismatched `quinn` versions, lol.

**Pull request merged**